### PR TITLE
add `Faraday#Deprecate` to `1.x`

### PR DESCRIPTION
**Pull request by hyuraku**
## Description
add Faraday#Deprecate to hide deprecate method's warning message

Fixes https://github.com/lostisland/faraday/issues/1410

## Todos
List any remaining work that needs to be done, i.e:
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: lib/faraday/deprecate.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  #     end
+  module Deprecate
+    def self.skip # :nodoc:
+      @skip ||= begin
+        case ENV['FARADAY_DEPRECATE'].to_s.downcase
+        when '1', 'warn' then :warn
+        else :skip
+        end
+      end
+      @skip = :warn
```
Comment (iMacTia) [inline lib/faraday/deprecate.rb:57]: This line looks like a mistake 🤔

File: lib/faraday/deprecate.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    end
+
+    def self.skip=(value) # :nodoc:
+      @skip = value ? :skip : :warn
+    end
+
+    # Temporarily turn off warnings. Intended for tests only.
+    def skip_during
+      original = Faraday::Deprecate.skip
+      Faraday::Deprecate.skip, = true
```
Comment (iMacTia) [inline lib/faraday/deprecate.rb:68]: There's an extra `,` here (I noticed this was also in the original commit! But should probably remove it now)

**Review by iMacTia — CHANGES_REQUESTED** [review]
Thank you for bringing this back @hyuraku.
There are some changes that I wouldn't expect in this PR, please see my comment below.
Do you also think you could add some test coverage to check the warnings are correctly output when the skip is disabled?

File: lib/faraday/deprecate.rb (hunk 1 above)
Comment (hyuraku) [reply lib/faraday/deprecate.rb:57]: This is a code for debugging. I remove it soon.

**Comment by iMacTia** [comment]
Thank you for addressing all the comments @hyuraku 🙌!
Are you planning to add unit tests as well as I mentioned above?

> Do you also think you could add some test coverage to check the warnings are correctly output when the skip is disabled?

**Comment by hyuraku** [comment]
@iMacTia 
> check the warnings are correctly output when the skip is disabled?

I will add the test when the skip is disabled? to `spec/faraday/connection_spec.rb`

File: lib/faraday/error.rb [hunk 3]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'faraday/deprecate'
```
Comment (iMacTia) [inline lib/faraday/error.rb:3]: This is not necessary anymore 🙌 !

**Comment by iMacTia** [comment]
Or better, a generic test around the `Faraday::Deprecate#deprecate` method 👍!
If we know that works as expected (e.g. with an anonymous class defined in the specs), we don't need to test it everywhere we use it

**Review by iMacTia — COMMENTED** [review]
Just realised we can now remove the `require` from `error.rb` as well

**Comment by hyuraku** [comment]
@iMacTia 
I added the spec for `Faraday::Deprecate#deprecate`; could you review this PR?

**Review by iMacTia — APPROVED** [review]
Fantastic work @hyuraku and great test coverage 💯👏
The `rubocop` offence is unrelated to the PR itself, I'll test this locally and merge if all tests are passing 👍

**Comment by iMacTia** [comment]
All tests pass locally ✅ 🎉

**Comment by kyrofa** [comment]
Thank you both!

**Review by kenichi-sonoda — COMMENTED** [review]
Sorry, mistake

**Pull request merged**